### DISCUSSION TITLE
Use-Case for synchronizing a dynamic subset of documents

### DISCUSSION THREAD

**Discussion by gruuna** [cofu-app/cbl-dart#500] [General]
Following scenario:

We have a database Server with REST-methods which holds customer data and we have a flutter app that wants to replicate some of these customers to its local database.

At the moment, the app has 2 modes:
- "online": it queries all customers directly by REST-requests from the server and caches them in local database
- "offline": it queries only locally available customers which are "synchronized" manually (user selects customer to synchronize)

Now we want to switch to couchbase (and cbl-dart) to use it's great replication functionality... But we don't know how to solve our scenario from above where a user can select the customers that have to be replicated...

In other words: user must be able to "browse" available customers online and then select the customer (-ids) to replicate locally.

Our first approach was to make the IDs "searchable" like "12345-customerName-customerAddress-..." to use these keys in replicator-config "List<String>? documentIds" but we don't think that this is good document-design isn't it?

Thanks in advance!

**Comment by blaugold** [comment]
Hey @gruuna. This sounds like a use-case for channels. You can configure the sync function in the Couchbase Sync Gateway to assign documents to channels based on the data contained within each document. Then you can configure a `Replicator` to only replicate certain channels. In your case, the documents belonging to one customer would all be assigned to a channel for that customer.

**Reply by gruuna** [reply]
Hi @blaugold and thanks for your answer! Yes, channels would also be an option - but we need the customers BEFORE they're assigned to any channel...

Other approach:
- User "browses/searches" for any customers with customer-id beginning with "A-*" 
- couchbase (Sync Gateway? or Server?) gives the search-results (documents with IDs)
- User selects customer ID "A-12345" 
- Replicator starts replication of customer "A-12345"

BUT: how to realize this "search" or "browse"? Because the sync-gateway has no option to search documents and when we have to use the couchbase server, our app relies on 2 different endpoints (sync gateway and server).

**Reply by blaugold** [reply]
Ah, ok. I think I understand. You can implement a REST endpoint for searching / browsing customers by accessing Couchbase Server with the help of one of the available Couchbase Server SDKs. There are SDKs available [for most languages](https://docs.couchbase.com/home/sdk.html) that are commonly used in backends.

You could use an SQL++ query similar to this one to only select the data that is relevant for browsing and page through it:

```sql
SELECT META().id, name, address
FROM `users`
ORDER BY name
LIMIT 25
OFFSET 100
```

[Here is for example](https://docs.couchbase.com/nodejs-sdk/current/howtos/n1ql-queries-with-sdk.html) is how to query a Couchbase Server from a Node.JS app.

**Reply by gruuna** [reply]
@blaugold So, do you say that it's a common approach to use 2 endpoints: 
- couchbase server for remote-querying and searching
- sync gateway (with cbl-dart) for sync and all local things
?

**Reply by blaugold** [reply]
Yes, if you have large amounts of data that cannot be stored locally in its entirety that's a common approach.  

Another option that might be possible is to maintain small summary documents for each customer document. If there aren't  too many customer documents those summaries could all be assigned to one channel that is synced by all clients. Then you can implement searching and browsing locally as well.
